src/compile/ast.rs

```diff
@@ -712,6 +925,32 @@ pub fn type_vectors_assignable(
             .all(|(t1, t2)| t1.assignable(t2, type_tree, seen.clone()))
 }
 
+fn field_vectors_downcastable(
+    tvec1: &[StructField],
+    tvec2: &[StructField],
+    type_tree: &TypeTree,
+    seen: HashSet<(Type, Type)>,
+) -> bool {
+    tvec1.len() == tvec2.len()
+        && tvec1
+            .iter()
+            .zip(tvec2)
+            .all(|(t1, t2)| t1.tipe.downcastable(&t2.tipe, type_tree, seen.clone()))
+}
+
+fn field_vectors_castable(
+    tvec1: &[StructField],
+    tvec2: &[StructField],
+    type_tree: &TypeTree,
+    seen: HashSet<(Type, Type)>,
+) -> bool {
+    tvec1.len() == tvec2.len()
+        && tvec1
+            .iter()
+            .zip(tvec2)
+            .all(|(t1, t2)| t1.tipe.castable(&t2.tipe, type_tree, seen.clone()))
```

Comment:
Should we insist that the field names match here?  That seems consistent with our working definition of castable.

Reply:
I am unclear on what exactly you mean by "our working definition of castable", if you mean our current definition of `assignability`, that is replicated by `weakcast`, if you are talking about the definition proposed in #477 , this behavior matches the proposal.